langchain4j-core/src/main/java/dev/langchain4j/model/input/structured/DefaultStructuredPromptFactory.java

```diff
@@ -0,0 +1,51 @@
+package dev.langchain4j.model.input.structured;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.ToNumberPolicy;
+import com.google.gson.reflect.TypeToken;
+import dev.langchain4j.model.input.Prompt;
+import dev.langchain4j.model.input.PromptTemplate;
+import dev.langchain4j.spi.prompt.structured.StructuredPromptFactory;
+import java.util.Map;
+
+public class DefaultStructuredPromptFactory implements StructuredPromptFactory {
+    private static final Gson GSON = new GsonBuilder().setObjectToNumberStrategy(ToNumberPolicy.LONG_OR_DOUBLE).create();
+
+    public Prompt toPrompt(Object structuredPrompt) {
+        validate(structuredPrompt);
+
+        StructuredPrompt annotation = structuredPrompt.getClass().getAnnotation(StructuredPrompt.class);
+
+        String promptTemplateString = String.join(annotation.delimiter(), annotation.value());
+        PromptTemplate promptTemplate = PromptTemplate.from(promptTemplateString);
+
+        Map<String, Object> variables = extractVariables(structuredPrompt);
+
+        return promptTemplate.apply(variables);
+    }
+
+    private void validate(Object structuredPrompt) {
+        if (structuredPrompt == null) {
+            throw new IllegalArgumentException("Structured prompt cannot be null");
+        }
+
+        String structuredPromptClassName = structuredPrompt.getClass().getName();
+
+        StructuredPrompt structuredPromptAnnotation = structuredPrompt.getClass().getAnnotation(StructuredPrompt.class);
+        if (structuredPromptAnnotation == null) {
+            throw new IllegalArgumentException(
+                    String.format(
+                            "%s should be annotated with @StructuredPrompt to be used as a structured prompt",
+                            structuredPromptClassName
+                    )
+            );
+        }
+    }
+
+    private static Map<String, Object> extractVariables(Object structuredPrompt) {
```

Comment:
The functionality of this class should remain the same for other implementations, the difference is how the metadata is obtained.

In my Quarkus implementation for example, there is no need to do look up any annotations or do any validation because that functionality is provided at build time.